### PR TITLE
Untangle Compilation object from BSP client

### PR DESCRIPTION
**Pull request by odisseus**
This pull request is related to issues #987 and #1077.

### CODE REVIEW

File: src/compile/compile.scala [hunk 1]
```diff
@@ -211,6 +211,8 @@ object Compilation {
 
   private val compilationCache: collection.mutable.Map[Path, Future[Try[Compilation]]] = TrieMap()
 
+  val requestOrigins: collection.mutable.Map[String, Compilation] = TrieMap()
```
Comment (propensive) [inline src/compile/compile.scala]: Can we make this private?

File: src/compile/compile.scala [hunk 2]
```diff
@@ -146,7 +146,7 @@ object BloopServer extends Lifecycle.Shutdown with Lifecycle.ResourceHolder {
 
       thread.start()
       
-      val client = new FuryBuildClient(compilation, targetId, layout)
+      val client = new FuryBuildClient(targetId, layout)
```
Comment (propensive) [inline src/compile/compile.scala:148]: Great!

File: src/compile/compile.scala [hunk 3]
```diff
@@ -285,21 +286,26 @@ object Compilation {
   }
 
   private val originIdCounter = new AtomicInteger(1)
-  def nextOriginId()(implicit log: Log): String = {
-    s"${log.pid.pid}-${originIdCounter.getAndIncrement}"
+
+  def nextOriginId(compilation: Compilation)(implicit log: Log): String = {
+    val originId = s"${log.pid.pid}-${originIdCounter.getAndIncrement}"
```
Comment (propensive) [inline src/compile/compile.scala]: Is there a reason why we're using a stringly-typed index for the origin ID? Can you create an `OriginId` type in `ids.scala`?

File: src/compile/compile.scala [hunk 4]
```diff
@@ -285,21 +286,26 @@ object Compilation {
...
 
-  private def broadcast(event: CompileEvent): Unit = {
+  def broadcast(event: CompileEvent): Unit = {
     val ref = event match {
       case e: ModuleCompileEvent => e.ref
       case _ => targetId.ref
     }
-    BloopServer.subscribers(this).map(_.multiplexer).foreach(_(ref) = event)
+    BloopServer.subscribers(this).map(_.multiplexer)
+      .filter(_.contains(ref)).foreach(_(ref) = event)
```
Comment (propensive) [inline src/compile/compile.scala:307]: It's probably fine, but can you check whether `contains` (on whatever `multiplexer` is) is constant time, i.e. that it's not linear for every single event we receive (and hence _O(nm)_).

File: src/compile/compile.scala [hunk 5]
```diff
@@ -389,22 +396,32 @@ ${'|'} ${highlightedLine}
...
       val targetId = getCompileTargetId(params.getData)
       val ref = targetId.ref
-      broadcast(StopCompile(ref, params.getStatus == StatusCode.OK))
-      if(!compilation.targets(ref).kind.needsExecution) broadcast(StopRun(ref))
-      else broadcast(StartRun(ref))
+      broadcast( StopCompile(ref, params.getStatus == StatusCode.OK))
+      val compilations = Compilation.requestOrigins.values.filter(_.target.id == targetId)
+      for {
+        compilation <- compilations
+      } yield {
```
Comment (propensive) [inline src/compile/compile.scala]: I think, `compilations.foreach { ... }` makes it clearer that this returns `Unit`.

File: src/compile/compile.scala (hunk 4 above)
Comment (propensive) [reply src/compile/compile.scala:307]: Ok, I saw your implementation below. It defers to `keys` in the multiplexer which is a `List`, so it's linear. Can we make `keys` a `Set`?

**Review by propensive — CHANGES_REQUESTED** [review]

File: src/compile/compile.scala (hunk 4 above)
Comment (odisseus) [reply src/compile/compile.scala:307]: Good point

File: src/compile/compile.scala (hunk 5 above)
Comment (odisseus) [reply src/compile/compile.scala]: Refactored

File: src/compile/compile.scala (hunk 1 above)
Comment (odisseus) [reply src/compile/compile.scala]: Refactored

**Review by propensive — APPROVED** [review]

**Pull request merged**